node/src/isolate/external.cpp

```diff
@@ -0,0 +1,200 @@
+    }
+
+    void on_done() {
+        spool_handle->on_ready();
+    }
+
+    void on_error(std::error_code ec, const std::string& msg) {
+        spool_handle->on_abort(ec, msg);
+    }
+
+    std::shared_ptr<api::spool_handle_base_t> spool_handle;
+
+};
+
+struct spool_cancellation_t :
+    public api::cancellation_t
+{
+    spool_cancellation_t(io::upstream_ptr_t _stream) :
+        stream(std::move(_stream))
+    {}
+
+    ~spool_cancellation_t() {
+        cancel();
+    }
+
+    virtual
+    void
+    cancel() {
+        if(cancelled.test_and_set()) {
+            stream->send<io::isolate_spooled::cancel>();
+        }
+    }
+
+    io::upstream_ptr_t stream;
+    std::atomic_flag cancelled;
+};
+
+struct spawn_dispatch_t :
+    public dispatch<io::event_traits<io::isolate::spawn>::upstream_type>
+{
+    typedef io::protocol<io::event_traits<io::isolate::spawn>::upstream_type>::scope protocol;
+
+    spawn_dispatch_t(const std::string& name, std::shared_ptr<api::spawn_handle_base_t> _spool_handle):
+        dispatch(name),
+        spawn_handle(std::move(_spool_handle))
+    {
+        on<protocol::chunk>(std::bind(&spawn_dispatch_t::on_chunk, this, std::placeholders::_1));
+        on<protocol::choke>(std::bind(&spawn_dispatch_t::on_choke, this));
+        on<protocol::error>(std::bind(&spawn_dispatch_t::on_error, this, std::placeholders::_1, std::placeholders::_2));
+    }
+
+    void on_chunk(std::string chunk) {
+        if(chunk.empty()) {
+            spawn_handle->on_ready();
+        } else {
+            spawn_handle->on_data(std::move(chunk));
+        }
+
+    }
+
+    void on_choke() {
+        spawn_handle->on_terminate(std::error_code(), "");
+    }
+
+    void on_error(std::error_code ec, const std::string& msg) {
+        spawn_handle->on_terminate(ec, msg);
+    }
+
+    std::shared_ptr<api::spawn_handle_base_t> spawn_handle;
+};
+
+struct spawn_cancellation_t :
+public api::cancellation_t
+{
+    spawn_cancellation_t(io::upstream_ptr_t _stream) :
+        stream(std::move(_stream))
+    {}
+
+    ~spawn_cancellation_t() {
+        cancel();
+    }
+
+    virtual
+    void
+    cancel() {
+        if(cancelled.test_and_set()) {
+            stream->send<io::isolate_spawned::kill>();
+        }
+    }
+
+    io::upstream_ptr_t stream;
+    std::atomic_flag cancelled;
+};
+
+struct external_t::inner_t :
+    public std::enable_shared_from_this<inner_t>
+{
+    context_t& context;
+    asio::io_service& io_context;
+    std::unique_ptr<asio::ip::tcp::socket> socket;
+    std::shared_ptr<session_t> session;
+    const std::string name;
+    const dynamic_t args;
+    std::unique_ptr<logging::logger_t> log;
+
+    inner_t(context_t& _context, asio::io_service& _io_context, const std::string&, const dynamic_t& _args) :
+        context(_context),
+        io_context(_io_context),
+        socket(new asio::ip::tcp::socket(io_context)),
```

Comment:
Don't do this please. Sockets can't be reused after reaching terminated state - they are required to be recreated each time before connecting.